File: javascript/leetcode/subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} s
 * @param {TreeNode} t
 * @return {boolean}
 */
var isSubtree = function(s, t) {
    return traverse(s, t);
}

var isEqual = function(s, t) {
    if (s == null && t == null) {
        return true
    }
    if (s == null || t == null) {
        return false
    }
    return s.val == t.val && isEqual(s.left, t.left) && isEqual(s.right, t.right)
};

var traverse = function (s, t) {
    return s != null && (isEqual(s, t) || traverse(s.left, t) || traverse(s.right, t))
};

File: java/leetcode/subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode s, TreeNode t) {
        return traverse(s, t);
    }
    
    private boolean isEqual (TreeNode s, TreeNode t) {
        if (s == null && t == null) {
            return true;
        }
        if (s == null || t == null) {
            return false;
        }
        return s.val == t.val && isEqual(s.left, t.left) && isEqual(s.right, t.right);
    }
    
    private boolean traverse (TreeNode s, TreeNode t) {
        return s != null && (isEqual(s, t) || traverse(s.left, t) || traverse(s.right, t));
    }
}
